src/interactivity/base/HostSignalInputThread.cpp

```diff
@@ -0,0 +1,255 @@
+                }
+
+                LOG_IF_NTSTATUS_FAILED(ServiceLocator::LocateConsoleControl()->EndTask(ULongToHandle(msg.dwProcessId), msg.dwEventType, msg.ulCtrlFlags));
+            }
+            else
+            {
+                return E_ABORT;
+            }
+            break;
+        }
+        default:
+        {
+            THROW_HR(E_UNEXPECTED);
+            break;
+        }
+        }
+    }
+    return S_OK;
+}
+
+// Method Description:
+// - Skips the file stream forward by the specified number of bytes.zs
+// Arguments:
+// - cbBytes - Count of bytes to skip forward
+// Return Value:
+// - True if we could skip forward successfully. False otherwise.
+bool HostSignalInputThread::_AdvanceReader(const DWORD cbBytes)
+{
+    std::array<BYTE, 8> buffer;
+    auto bytesRemaining = cbBytes;
+
+    while (bytesRemaining > 0)
+    {
+        const auto advance = std::min(bytesRemaining, gsl::narrow_cast<DWORD>(buffer.max_size()));
+
+        if (!_GetData(buffer.data(), advance))
+        {
+            return false;
+        }
+
+        bytesRemaining -= advance;
+    }
+
+    return true;
+}
+
+// Method Description:
+// - Retrieves bytes from the file stream and exits or throws errors should the pipe state
+//   be compromised.
+// Arguments:
+// - pBuffer - Buffer to fill with data.
+// - cbBuffer - Count of bytes in the given buffer.
+// Return Value:
+// - True if data was retrieved successfully. False otherwise.
+bool HostSignalInputThread::_GetData(_Out_writes_bytes_(cbBuffer) void* const pBuffer,
+                                     const DWORD cbBuffer)
+{
+    DWORD dwRead = 0;
+    // If we failed to read because the terminal broke our pipe (usually due
+    //      to dying itself), close gracefully with ERROR_BROKEN_PIPE.
+    // Otherwise throw an exception. ERROR_BROKEN_PIPE is the only case that
+    //       we want to gracefully close in.
+    if (FALSE == ReadFile(_hFile.get(), pBuffer, cbBuffer, &dwRead, nullptr))
+    {
+        DWORD lastError = GetLastError();
+        if (lastError == ERROR_BROKEN_PIPE)
+        {
+            _Shutdown();
+            return false;
+        }
+        else
+        {
+            THROW_WIN32(lastError);
+        }
+    }
+    else if (dwRead != cbBuffer)
+    {
+        _Shutdown();
+        return false;
+    }
+
+    return true;
+}
+
+// Method Description:
+// - Starts the PTY Signal input thread.
+[[nodiscard]] HRESULT HostSignalInputThread::Start() noexcept
+{
+    RETURN_LAST_ERROR_IF(!_hFile);
+
+    HANDLE hThread = nullptr;
+    // 0 is the right value, https://blogs.msdn.microsoft.com/oldnewthing/20040223-00/?p=40503
+    DWORD dwThreadId = 0;
+
+    hThread = CreateThread(nullptr,
+                           0,
+                           HostSignalInputThread::StaticThreadProc,
+                           this,
+                           0,
+                           &dwThreadId);
+
+    RETURN_LAST_ERROR_IF_NULL(hThread);
+    _hThread.reset(hThread);

```

Comment:
You can probably just store the output of CreateThread directly into `_hThread`, no?

Reply:
Same for `_dwThreadId`.

It's safe from torn state because you're torching the house if it fails.